### ISSUE TITLE
Skip parts when materializing TTL for TTL-expressions that are monotonic functions of columns

### ISSUE THREAD

**Issue by siradjev** [ClickHouse/ClickHouse#20451]
**Problem description**
There is some fundamental problem with current TTL methodology that prevents from using it in production-grade non-cloud environment (when quick elastic resource allocation is not possible) for automated data retention purpose without impact to ongoing operation.  
Whenever TTL expression(s) is (are) modified or "MATERIALIZE TTL" is called manually all the parts have to mutate to recalculate expression value. (https://github.com/ClickHouse/ClickHouse/issues/10128#issuecomment-769782785)
Even though there is no need to touch 90% of parts all of them are mutated (it may be with hardlinking, but it's still heavy operation if there are many columns x parts). 
This directly brings down DB for quite a while by creating simply too many mutation entries in replication queue causing picture as below:
* merges are not executed due to "number_of_free_entries_in_pool_to_lower_max_size_of_merge":
`Not executing log entry MERGE_PARTS for part xxxxx because source parts size (481.19 MiB) is greater than the current maximum (4.45 MiB).`
* mutations just start failing with "number_of_free_entries_in_pool_to_execute_mutation":
`Not executing log entry MUTATE_PART for part xxxxx because source parts size (27.52 KiB) is greater than the current maximum (0.00 B).`

That one is probably problem on its own as free pool is checked after getting from queue, and there is always no free space in pool when queue is full. It's resolved after manually changing number_of_free_entries_in_pool_to_execute_mutation=1/0, so that clickhouse does not skips execution of log entries. 

Improvements listed in #10128 under "A large number of background TTL operations..." are not sufficient, as they address different aspect of this problem. 

**Typical use cases of using TTL**
* whole table data retention or large column data retention - typical expression is `timestampCol + interval`
* moving cold data to slower storage tiers - typical expression is `timestampCol + interval`

**High-level functional requirement**
* Adding/modifying TTL expression that is a monotonic function of one or several of partitionkey, or primarykey columns, MUST NOT cause any modifications/mutations of existing parts. Modifications of TTL are normal routine when initial dimensioning is not done. Vast majority of usages in community and clickhouse manual is based on `partitionkey + interval` kind of expression. So this alone shall provide huge improvement
* Modifying TTL expression that is a monotonic function of set of columns when set of columns is not changed, MUST NOT cause any modifications/mutations of existing parts.  E.g. if we had `TTL myDate+INTERVAL 3 day` it should not cause any change when we modify it as `TTL myDate+INTERVAL 4 DAY`. It is normal case to increase/decrease number of days to address input data stream growth. 
* materialize_ttl_after_modify default MUST be 0 to prevent huge overhead of reworking of old parts for the remaining cases. If required `ALTER TABLE xxx MODIFY TTL MATERIALIZE` can be used to trigger it. Having it as default 1 in 20.8 (#11042) was simple introduction of dangerous non-backward compatible change without feasible way of "undoing" it. 

**Describe the solution you'd like**
When expression is monotonic function of column(s), columns min/max themselves are enough. They are already kept for partition key, primary key, so when ttl is added on one of these there is no need to do anything. 
When ttl_only_drop_parts=0 parts that partially satisfy TTL expression will be mutated, when ttl_only_drop_parts=1 only parts that entirely satisfy TTL expression will be mutated. 
If expression has other columns - their minmax will also be required to be kept. 
If one of the columns used in TTL is removed from expression - there is no need to mutate all parts. 
The only cases when mutation on all parts are really needed:
* expression is non-monotonic to columns used in it
* expression is modified with a new column added that is not partition key and that is not a primary key
In all other cases parts that don't satisfy TTL expression will be simply skipped after run of MATERIALIZE TTL. 

**Describe alternatives you've considered**
External shell scripting.

**Comment by filimonov** [comment]
Fixing of https://github.com/ClickHouse/ClickHouse/issues/20338 should make the mutations very cheap/close to instant (as they should be). 

About limiting mutations to the scope of partitions - there is a task 
https://github.com/ClickHouse/ClickHouse/issues/12188

It was mplemented partially already (via ALTER TABLE ... IN PARTITION ... syntax for ReplicatedMergeTree) - see https://github.com/ClickHouse/ClickHouse/pull/13403

**Comment by siradjev** [comment]
Thank you for your feedback. 

> Fixing of #20338 should make the mutations very cheap/close to instant (as they should be).
With high number of parts (>1k) it's still an issue, especially in replicated MTs. In case of this improvement it will help minimize impact of increase/decrease of data retentions or configuring data retention based on partition key (which is most frequent reason for TTL modification). 

> About limiting mutations to the scope of partitions - there is a task #12188
I think this one is more about "ALTER TABLE xxx WHERE cond" case to make partition pruning. 
It won't help with TTL modification as TTLs are modified globally on table/column level, not on part level.

**Comment by xjewer** [comment]
>When expression is monotonic function of column(s), columns min/max themselves are enough. They are already kept for partition key, primary key, so when ttl is added on one of these there is no need to do anything.
When ttl_only_drop_parts=0 parts that partially satisfy TTL expression will be mutated, when ttl_only_drop_parts=1 only parts that entirely satisfy TTL expression will be mutated.
If expression has other columns - their minmax will also be required to be kept.
If one of the columns used in TTL is removed from expression - there is no need to mutate all parts.
The only cases when mutation on all parts are really needed:
>* expression is non-monotonic to columns used in it
>* expression is modified with a new column added that is not partition key and that is not a primary key
In all other cases parts that don't satisfy TTL expression will be simply skipped after run of MATERIALIZE TTL.

literally recently discussed the same approach internally. This optimisation does make sense, though would break backward compatibility, since optimised case won't require having TTL defined per part.

**Comment by siradjev** [comment]
>This optimisation does make sense, though would break backward compatibility, since optimised case won't require having TTL defined per part.

It does not have to break backward compatibility as old format parts can still be usable with newer version of clickhouse that implements this improvement.

**Comment by xjewer** [comment]
>It does not have to break backward compatibility as old format parts will still be usable with newer version of clickhouse that implements this improvement.

that's forward, not backward.

I mean the use case, where optimisation took place on a new CH version (table metadata was changed, but parts weren't mutated), and then after downgrading to an older version -> TTL won't work as expected.

**Comment by siradjev** [comment]
> I mean the use case, where optimisation took place on a new CH version (table metadata was changed, but parts weren't mutated), and then after downgrading to an older version -> TTL won't work as expected.

I think there is versioning supported as well in those text files in the parts information so after rollback, older version can keep using old format. So older version would keep working as is.

**Comment by siradjev** [comment]
> Fixing of #20338 should make the mutations very cheap/close to instant (as they should be).

This is correct, but for the case of many partitions it would still cause a lot of mutations and in case of replicated merge tree those are really bad in terms of maintenance. 

> About limiting mutations to the scope of partitions - there is a task #12188
> It was mplemented partially already (via ALTER TABLE ... IN PARTITION ... syntax for ReplicatedMergeTree) - see #13403

@filimonov, I think those are not about TTL modification as it's table level configuration, and it can't be applied on SOME partitions. TTL modification is actually quite a frequent activity. Implementing this issue can help to avoid mutation on those changes altogether.

**Comment by nvartolomei** [comment]
I'm interested in this feature and spent some time looking at how it could/should be implemented. So far, the task for adding optimization for skipping reading entire columns seems complicated. The complications arise from the flexibility of the TTL feature in ClickHouse. That is that it accept a wide variety of expressions which are hard to be proven monotonic. For example even the simplest one `date + interval 31 day` is hard (impossible) to prove as monotonic on metadata level as `plus` operation can overflow.

I was thinking about _carefully_ analyzing the TTL expression and if it is of the form `always_monotonic_expr + constant` then apply the above suggested optimizaton on source column minmax values.

Another alternative is to introduce a new syntax that accepts only this form of TTLs (`OLDER_THAN(_column_, _ttl_)`). Btw, this approach is similar to the one found in Google Spanner https://cloud.google.com/spanner/docs/ttl/working-with-ttl#syntax, though I can't say if they came up with this one for the same reason.

This new syntax may seem more constrained than the current one, but it is not. Spanner docs suggest to use a derived column in TTLs if you need complex logic for deleting data.

```
CREATE TABLE OrderTable (
  OrderId INT64 NOT NULL,
  OrderStatus STRING(30) NOT NULL,
  LastModifiedDate TIMESTAMP NOT NULL,
  ExpiredDate TIMESTAMP AS (IF(OrderStatus = "Cancelled",
    TIMESTAMP_ADD(LastModifiedDate, INTERVAL 30 DAY),
    TIMESTAMP_ADD(LastModifiedDate, INTERVAL 180 DAY))) STORED,
) PRIMARY KEY(OrderId)
, ROW DELETION POLICY (OLDER_THAN(ExpiredDate, INTERVAL 0 DAY));
```

The beauty of this approach is that users can expect that updating a column is a "heavy" task. We wouldn't have to communicate that oh we have lightweight TTL expressions in some conditions and heavy TTL updates in others.

It also seems to me that with this approach we could avoid the overflow problem altogether.

TTLs on columns used in partition expression become metadata only. If ClickHouse decides to maintain a minmax index for all columns then any TTL becomes metadata only.
